docs/proposal-001_add-ons.md

```diff
@@ -0,0 +1,151 @@
+# Desing Proposal #001: Add-ons
+
+## What's an add-ons?
+
+An add-on extends functionality of a Kubernetes cluster, it may consist of a workload and/or configuration within the give cluster or the cloud provider. The workload, if present, would may classify as an operator or a controller, however that is not neccessary.
+
+The difference between add-ons and anything a user decided to run themselves, is that from user's point of view, add-on is something they don't need to maintain directly. A cluster provider or bootstrap tool (i.e. eksctl), is expected to ensure ease of use, by providing minumum viable (that configuration to ensure end-to-end integration and compatibility), as well as catering for reconfiguration and upgrades.
+
+## Design Requirements and Goals
+
+Add-on must expose all of the following attributes:
+
+- name (fixed) *[TBD]*
+- version tag or revision (has default value that user can override)
+- source URL (has default value that user can override)
+- set of customisations (no default value, user defined)
+
+Internally, an add-on may consist of any and/or all of the following elements:
+
+- Kubernetes workload definitions
+- Kubernetes configuration objects - PersistentVolumes, ConfigMaps, ...etc
+- cloud provider resource definitons (e.g. CloudFormation sub-resources)
+- dependencies (supported Kubernetes versions and corresponding cloud providers, list of references to other add-ons)
```

Comment:
When we say `list of references to other add-ons` we mean to a specific version of that addon? If we do mean that, then will we support the NPM (or something else) style of specifiying versions (e.g. 1.2.1, >1.2.1, ~2.2.1, ^1.2.1)?

Reply:
Good question! I am not sure, perhaps to start with we could start 

I'd be challenging to support multiple versions of the same add-on, but hopefully that won't be required, especially considering that most of the add-ons provide cluster-wide functionality. Perhaps we should state as part of the definition at the very top that every add-on instance is a singleton. In terms of the API, we may allow multiple `AddonInstance` objects with different names representing effectively the same add-on, but it'd be up to the actual add-on to implement namespacings of  its own (e.g. Flux by default runs cluster-wide, but its responsibility can be limited to a namespace).